### PR TITLE
Remove empty Threadpool .cctors

### PR DESCRIPTION
**Pull request by benaadams**
Do `ThreadPoolGlobals`, `QueueUserWorkItemCallback` and `_ThreadPoolWaitOrTimerCallback` need empty static constructors?

### CODE REVIEW

**Comment by benaadams** [comment]
Need to swap in func, 1 sec

**Comment by stephentoub** [comment]
> Do ThreadPoolGlobals, QueueUserWorkItemCallback and _ThreadPoolWaitOrTimerCallback need empty static constructors?

I expect these are there because in a world of Security{Safe}Critical and SecurityTransparent, initializing a static SecurityCritical field can only be done from a Security{Safe}Critical static cctor, and to attribute it appropriately you need the explicit cctor.  For mscorlib in CoreCLR, I'm not sure if that matters.  @AlexGhiondea or @jkotas can probably comment more authoritatively.

**Comment by benaadams** [comment]
Ah, that would make sense.

**Comment by benaadams** [comment]
@stephentoub would something like this work? Have put the initializers in a `SecuritySafeCritical` path for user code (`QueueUserWorkItem`) which go via `QueueUserWorkItemHelper` to `EnsureVMInitialized`

**Comment by stephentoub** [comment]
@ericeil might be the best to comment.

**Comment by ericeil** [comment]
This seems reasonable to me.

**Comment by benaadams** [comment]
Rebased and squished

**Comment by AlexGhiondea** [comment]
I am not sure I understand the advantage of this change. 

Sure, we don't have a static constructor that does seems to do nothing, but this will make it way harder to understand the way transparency works in those cases. 
And I would argue that the static constructor does something: sets up he security model.

While for now we found the code that called it and marked it appropriately, i think that has a couple of problems:
- now the caller's transparency has changed so we might have to potentially update its callers
- in the future, we are going to have to ensure that each caller has the right attribute applied which is very easy to forget to do
- the compiler already generates the static constructor and combine the empty one with the one that does the field initializer.

Can you please explain what advantage this change has over leaving it as-is?

**Comment by stephentoub** [comment]
> Can you please explain what advantage this change has over leaving it as-is?

There is cost to having an explicit static cctor, in that the C# compiler will then not mark the type as beforefieldinit, and the backend compiler will be forced to add initialization checks to all such statics.  It's something static analysis rules have checked for: https://msdn.microsoft.com/en-us/library/ms182275.aspx

I don't know whether it makes a measurable difference in this particular case.

**Comment by benaadams** [comment]
There is a gotcha that the first function that is called and does field access has the initialization checks compiled into it; while any later function does not. Seeing if I can rearrange.

**Comment by jkotas** [comment]
This does not mattet. mscorlib is always expected to be precompiled using crossgen in production - all static access has this check in precompiled code.

**Comment by benaadams** [comment]
:sob: might as well close then

**Comment by jkotas** [comment]
I just meant that reordering does not matter. There may be still benefit on your change by avoiding the cctor check in static methods that do not access any fields - have not looked whether there are any.

**Comment by benaadams** [comment]
@jkotas can initialize the four delegates to non-capturing lambdas, drop some of the static functions and a lot of the casts?

That work in the native realm?

File: src/mscorlib/src/System/Threading/ThreadPool.cs [hunk 1]
```diff
@@ -1882,8 +1883,7 @@ private static void EnsureVMInitialized()
         [System.Security.SecuritySafeCritical]
         internal static void NotifyWorkItemProgress()
         {
-            if (!ThreadPoolGlobals.vmTpInitialized)
-                ThreadPool.InitializeVMTp(ref ThreadPoolGlobals.enableWorkerTracking);
+            EnsureVMInitialized();
             NotifyWorkItemProgressNative();
         }
```
Comment (benaadams) [inline src/mscorlib/src/System/Threading/ThreadPool.cs]: Was the behaviour correct to not set `ThreadPoolGlobals.vmTpInitialized = true;` at this point previously?

**Comment by benaadams** [comment]
Last change drops the auto.cctor entirely (which is different from the non- beforefieldinit .cctor) and may mean no runtime guards are needed on the static fields?

**Comment by benaadams** [comment]
From
cctor
To
no-cctor

**Comment by benaadams** [comment]
For jitted stack trace - I think all the guards are gone - stack trace in

`public static void Main(string[] args)`
```
--- I:\IllyriadGames.AgeOfAscent\Test\ThreadpoolTest\src\ThreadpoolTest\Program.cs 
    12:         {
00007FFE2AA41B92  push        rsi  
00007FFE2AA41B93  sub         rsp,40h  
00007FFE2AA41B97  mov         rbp,rsp  
00007FFE2AA41B9A  xor         eax,eax  
00007FFE2AA41B9C  mov         qword ptr [rbp+30h],rax  
00007FFE2AA41BA0  mov         qword ptr [rbp+38h],rax  
00007FFE2AA41BA4  mov         qword ptr [rbp+60h],rcx  
00007FFE2AA41BA8  cmp         dword ptr [7FFE2AB0EBA8h],0  
00007FFE2AA41BAF  je          00007FFE2AA41BB6  
00007FFE2AA41BB1  call        00007FFE8A124760  
00007FFE2AA41BB6  nop  
    13:             ThreadPool.QueueUserWorkItem(MyFunc, null);
00007FFE2AA41BB7  mov         rcx,7FFE2A97D670h  
00007FFE2AA41BC1  call        00007FFE8A4116C0  
00007FFE2AA41BC6  mov         qword ptr [rbp+30h],rax  
00007FFE2AA41BCA  mov         r8,7FFE2AA3F140h  
00007FFE2AA41BD4  mov         rcx,qword ptr [rbp+30h]  
00007FFE2AA41BD8  xor         edx,edx  
00007FFE2AA41BDA  mov         r9,2548CFC3510h  
00007FFE2AA41BE4  call        00007FFE2A709130  
00007FFE2AA41BE9  mov         rcx,qword ptr [rbp+30h]  
00007FFE2AA41BED  xor         edx,edx  
00007FFE2AA41BEF  call        00007FFE2A98C390  
00007FFE2AA41BF4  mov         dword ptr [rbp+2Ch],eax  
00007FFE2AA41BF7  nop  
    14:         }
00007FFE2AA41BF8  nop  
00007FFE2AA41BF9  lea         rsp,[rbp+40h]  
00007FFE2AA41BFD  pop         rsi  
00007FFE2AA41BFE  pop         rdi  
00007FFE2AA41BFF  pop         rbp  
00007FFE2AA41C00  ret  
--- No source file -------------------------------------------------------------
```
`private void CtorOpened(Object target, IntPtr methodPtr, IntPtr shuffleThunk)`?
`public static bool QueueUserWorkItem(`
```
--- I:\Work\GitHub\coreclr\src\mscorlib\src\System\Threading\ThreadPool.cs -----
  1627:             StackCrawlMark stackMark = StackCrawlMark.LookForMyCaller;
00007FFE2A98C392  in          al,dx  
00007FFE2A98C393  sub         byte ptr [rbx],dh  
00007FFE2A98C395  ror         byte ptr [rax-77h],44h  
00007FFE2A98C399  and         al,20h  
00007FFE2A98C39B  mov         dword ptr [rsp+20h],1  
  1628:             return QueueUserWorkItemHelper(callBack,state,ref stackMark,true);
00007FFE2A98C3A3  lea         r8,[rsp+20h]  
00007FFE2A98C3A8  mov         r9d,1  
00007FFE2A98C3AE  call        00007FFE2A88B5F8  
00007FFE2A98C3B3  nop  
00007FFE2A98C3B4  add         rsp,28h  
00007FFE2A98C3B8  ret  
--- No source file -------------------------------------------------------------
```
`private static bool QueueUserWorkItemHelper(WaitCallback callBack, `
```
--- I:\Work\GitHub\coreclr\src\mscorlib\src\System\Threading\ThreadPool.cs -----
  1658:             bool success =  true;
00007FFE2A98C3D2  push        rsi  
00007FFE2A98C3D3  sub         rsp,30h  
00007FFE2A98C3D7  lea         rbp,[rsp+40h]  
00007FFE2A98C3DC  mov         qword ptr [rsp+20h],rsp  
00007FFE2A98C3E1  mov         qword ptr [rbp+10h],rcx  
00007FFE2A98C3E5  mov         qword ptr [rbp+18h],rdx  
00007FFE2A98C3E9  mov         dword ptr [rbp+28h],r9d  
  1659: 
  1660:             if (callBack != null)
00007FFE2A98C3ED  mov         rax,qword ptr [rbp+10h]  
00007FFE2A98C3F1  test        rax,rax  
00007FFE2A98C3F4  je          00007FFE2A98C414  
  1661:             {
  1662:                         //The thread pool maintains a per-appdomain managed work queue.
  1663:                 //New thread pool entries are added in the managed queue.
  1664:                 //The VM is responsible for the actual growing/shrinking of 
  1665:                 //threads. 
  1666: 
  1667:                 EnsureVMInitialized();
00007FFE2A98C3F6  call        00007FFE2A88B668  
00007FFE2A98C3FB  nop  
00007FFE2A98C3FC  mov         rcx,qword ptr [rsp+20h]  
00007FFE2A98C401  call        00007FFE2A98C44E  
00007FFE2A98C406  nop  
  1691:             }
  1692:             return success;
00007FFE2A98C407  mov         eax,1  
00007FFE2A98C40C  lea         rsp,[rbp-10h]  
00007FFE2A98C410  pop         rsi  
00007FFE2A98C411  pop         rdi  
00007FFE2A98C412  pop         rbp  
00007FFE2A98C413  ret  
  1687:             }
  1688:             else
  1689:             {
  1690:                 throw new ArgumentNullException("WaitCallback");
00007FFE2A98C414  mov         rcx,7FFE2A749D30h  
00007FFE2A98C41E  call        00007FFE8A4116C0  
00007FFE2A98C423  mov         rsi,rax  
00007FFE2A98C426  mov         ecx,139C2h  
00007FFE2A98C42B  mov         rdx,7FFE2A566020h  
00007FFE2A98C435  call        00007FFE8A120890  
00007FFE2A98C43A  mov         rdx,rax  
00007FFE2A98C43D  mov         rcx,rsi  
00007FFE2A98C440  call        00007FFE2A6F2738  
00007FFE2A98C445  mov         rcx,rsi  
00007FFE2A98C448  call        00007FFE8A123060  
00007FFE2A98C44D  int         3  
  1658:             bool success =  true;
00007FFE2A98C44E  push        rbp  
00007FFE2A98C44F  push        rdi  
00007FFE2A98C450  push        rsi  
  1658:             bool success =  true;
00007FFE2A98C451  sub         rsp,30h  
00007FFE2A98C455  mov         rbp,qword ptr [rcx+20h]  
00007FFE2A98C459  mov         qword ptr [rsp+20h],rbp  
00007FFE2A98C45E  lea         rbp,[rbp+40h]  
  1674:                 finally
  1675:                 {
  1676:                     ExecutionContext context = compressStack && !ExecutionContext.IsFlowSuppressed() ?
  1677:                         ExecutionContext.Capture(ref stackMark, ExecutionContext.CaptureOptions.IgnoreSyncCtx | ExecutionContext.CaptureOptions.OptimizeDefaultCase) :
  1678:                         null;
00007FFE2A98C462  movzx       ecx,byte ptr [rbp+28h]  
00007FFE2A98C466  test        cl,cl  
00007FFE2A98C468  jne         00007FFE2A98C46E  
00007FFE2A98C46A  xor         esi,esi  
00007FFE2A98C46C  jmp         00007FFE2A98C4A6  
00007FFE2A98C46E  mov         rcx,7FFE2A573020h  
00007FFE2A98C478  mov         edx,54Ch  
00007FFE2A98C47D  call        00007FFE8A11F410  
00007FFE2A98C482  mov         rsi,qword ptr [rax+18h]  
00007FFE2A98C486  test        rsi,rsi  
00007FFE2A98C489  jne         00007FFE2A98C4A6  
00007FFE2A98C48B  mov         rcx,7FFE2A573020h  
00007FFE2A98C495  mov         edx,54Ch  
00007FFE2A98C49A  call        00007FFE8A411D80  
00007FFE2A98C49F  mov         rsi,qword ptr [rax+0E78h]  
00007FFE2A98C4A6  mov         rcx,7FFE2A573020h  
00007FFE2A98C4B0  mov         edx,54Ch  
00007FFE2A98C4B5  call        00007FFE8A411D80  
00007FFE2A98C4BA  cmp         qword ptr [rax+0E78h],rsi  
00007FFE2A98C4C1  je          00007FFE2A98C4FD  
  1679: 
  1680:                     IThreadPoolWorkItem tpcallBack = context == ExecutionContext.PreAllocatedDefault ?
  1681:                                  new QueueUserWorkItemCallbackDefaultContext(callBack, state) :
  1682:                                  (IThreadPoolWorkItem)new QueueUserWorkItemCallback(callBack, state, context);
00007FFE2A98C4C3  mov         rcx,7FFE2A97DB88h  
00007FFE2A98C4CD  call        00007FFE8A4116C0  
00007FFE2A98C4D2  mov         rdi,rax  
00007FFE2A98C4D5  lea         rcx,[rdi+8]  
00007FFE2A98C4D9  mov         rdx,qword ptr [rbp+10h]  
00007FFE2A98C4DD  call        00007FFE8A40F7F0  
00007FFE2A98C4E2  lea         rcx,[rdi+18h]  
00007FFE2A98C4E6  mov         rdx,qword ptr [rbp+18h]  
00007FFE2A98C4EA  call        00007FFE8A40F7F0  
00007FFE2A98C4EF  lea         rcx,[rdi+10h]  
00007FFE2A98C4F3  mov         rdx,rsi  
00007FFE2A98C4F6  call        00007FFE8A40F7F0  
00007FFE2A98C4FB  jmp         00007FFE2A98C529  
00007FFE2A98C4FD  mov         rcx,7FFE2A97D7C0h  
00007FFE2A98C507  call        00007FFE8A4116C0  
00007FFE2A98C50C  mov         rdi,rax  
00007FFE2A98C50F  lea         rcx,[rdi+8]  
00007FFE2A98C513  mov         rdx,qword ptr [rbp+10h]  
00007FFE2A98C517  call        00007FFE8A40F7F0  
00007FFE2A98C51C  lea         rcx,[rdi+10h]  
00007FFE2A98C520  mov         rdx,qword ptr [rbp+18h]  
00007FFE2A98C524  call        00007FFE8A40F7F0  
  1683: 
  1684:                     ThreadPoolGlobals.workQueue.Enqueue(tpcallBack, true);
00007FFE2A98C529  mov         rcx,7FFE2A573020h  
00007FFE2A98C533  call        00007FFE8A411DC0  
00007FFE2A98C538  mov         rcx,qword ptr [rax+0EA0h]  
00007FFE2A98C53F  ?? ?? 
```
`private static void EnsureVMInitialized()`
```
--- I:\Work\GitHub\coreclr\src\mscorlib\src\System\Threading\ThreadPool.cs -----
  1843:             if (!ThreadPoolGlobals.vmTpInitialized)
00007FFE2A98C582  sub         rsp,28h  
00007FFE2A98C586  lea         rbp,[rsp+30h]  
00007FFE2A98C58B  mov         rcx,7FFE2A573020h  
00007FFE2A98C595  call        00007FFE8A411D50  
00007FFE2A98C59A  mov         rsi,rax  
00007FFE2A98C59D  movzx       ecx,byte ptr [rsi+891h]  
00007FFE2A98C5A4  test        ecx,ecx  
00007FFE2A98C5A6  jne         00007FFE2A98C5CF  
  1844:             {
  1845:                 ThreadPool.InitializeVMTp(ref ThreadPoolGlobals.enableWorkerTracking);
00007FFE2A98C5A8  lea         rcx,[rsi+892h]  
00007FFE2A98C5AF  call        00007FFE2A98BFD0  
  1846: 
  1847:                 ThreadPoolGlobals.Initialize();
00007FFE2A98C5B4  call        00007FFE2A98BE80  
  1848:                 QueueUserWorkItemCallback.Initialize();
00007FFE2A98C5B9  call        00007FFE2A98BF48  
  1849:                 QueueUserWorkItemCallbackDefaultContext.Initialize();
00007FFE2A98C5BE  call        00007FFE2A98BE40  
  1850:                 _ThreadPoolWaitOrTimerCallback.Initialize();
00007FFE2A98C5C3  call        00007FFE2A98BF88  
  1851: 
  1852:                 ThreadPoolGlobals.vmTpInitialized = true;
00007FFE2A98C5C8  mov         byte ptr [rsi+891h],1  
00007FFE2A98C5CF  lea         rsp,[rbp-8]  
00007FFE2A98C5D3  pop         rsi  
00007FFE2A98C5D4  pop         rbp  
00007FFE2A98C5D5  ret  
--- No source file -------------------------------------------------------------
```

**Comment by benaadams** [comment]
Jitted stack trace out
`void IThreadPoolWorkItem.ExecuteWorkItem()`
```
--- I:\Work\GitHub\coreclr\src\mscorlib\src\System\Threading\ThreadPool.cs -----
  1290:             ExecutionContext.Run(ExecutionContext.PreAllocatedDefault, ccb, this, true);
00007FFE2A98E9B0  push        rdi  
00007FFE2A98E9B1  push        rsi  
00007FFE2A98E9B2  sub         rsp,28h  
00007FFE2A98E9B6  mov         rsi,rcx  
00007FFE2A98E9B9  mov         rcx,7FFE2A573020h  
00007FFE2A98E9C3  mov         edx,54Ch  
00007FFE2A98E9C8  call        00007FFE8A411D80  
00007FFE2A98E9CD  mov         rdi,qword ptr [rax+0E78h]  
00007FFE2A98E9D4  mov         rcx,7FFE2A573020h  
00007FFE2A98E9DE  call        00007FFE8A411DC0  
00007FFE2A98E9E3  mov         rdx,qword ptr [rax+10B0h]  
00007FFE2A98E9EA  mov         rcx,rdi  
00007FFE2A98E9ED  mov         r8,rsi  
00007FFE2A98E9F0  call        00007FFE2A88E1C0  
00007FFE2A98E9F5  nop  
00007FFE2A98E9F6  add         rsp,28h  
00007FFE2A98E9FA  pop         rsi  
00007FFE2A98E9FB  pop         rdi  
00007FFE2A98E9FC  ret  
--- No source file -------------------------------------------------------------
```
`public static void Run(ExecutionContext executionContext, ContextCallback`
```
--- I:\Work\GitHub\coreclr\src\mscorlib\src\System\Threading\ExecutionContext.cs 
    87:             ExecutionContextSwitcher ecsw = default(ExecutionContextSwitcher);
00007FFE2A98EA22  push        rsi  
00007FFE2A98EA23  push        rdi  
00007FFE2A98EA24  push        rsi  
00007FFE2A98EA25  push        rbx  
00007FFE2A98EA26  sub         rsp,40h  
00007FFE2A98EA2A  lea         rbp,[rsp+60h]  
00007FFE2A98EA2F  xor         eax,eax  
00007FFE2A98EA31  mov         qword ptr [rbp-30h],rax  
00007FFE2A98EA35  mov         qword ptr [rbp-28h],rax  
00007FFE2A98EA39  mov         qword ptr [rsp+20h],rsp  
00007FFE2A98EA3E  mov         rsi,rcx  
00007FFE2A98EA41  mov         rdi,rdx  
00007FFE2A98EA44  mov         rbx,r8  
00007FFE2A98EA47  xor         ecx,ecx  
00007FFE2A98EA49  mov         qword ptr [rbp-30h],rcx  
00007FFE2A98EA4D  mov         qword ptr [rbp-28h],rcx  
    88:             try
    89:             {
    90:                 EstablishCopyOnWriteScope(ref ecsw);
00007FFE2A98EA51  mov         rcx,7FFE2A573020h  
00007FFE2A98EA5B  mov         edx,54Ch  
00007FFE2A98EA60  call        00007FFE8A11F410  
00007FFE2A98EA65  mov         r14,qword ptr [rax+18h]  
00007FFE2A98EA69  test        r14,r14  
00007FFE2A98EA6C  jne         00007FFE2A98EA89  
00007FFE2A98EA6E  mov         rcx,7FFE2A573020h  
00007FFE2A98EA78  mov         edx,54Ch  
00007FFE2A98EA7D  call        00007FFE8A411D80  
00007FFE2A98EA82  mov         r14,qword ptr [rax+0E78h]  
00007FFE2A98EA89  mov         qword ptr [rbp-30h],r14  
00007FFE2A98EA8D  call        00007FFE2A98D8A0  
00007FFE2A98EA92  mov         qword ptr [rbp-28h],rax  
    91: 
    92:                 ExecutionContext.Restore(executionContext);
00007FFE2A98EA96  mov         rcx,rsi  
00007FFE2A98EA99  call        00007FFE2A88E1C8  
    93:                 callback(state);
00007FFE2A98EA9E  mov         rcx,qword ptr [rdi+8]  
00007FFE2A98EAA2  mov         rdx,rbx  
00007FFE2A98EAA5  call        qword ptr [rdi+18h]  
00007FFE2A98EAA8  nop  
   103:             }
   104:             ecsw.Undo();
00007FFE2A98EAA9  mov         rsi,qword ptr [rbp-28h]  
00007FFE2A98EAAD  mov         rcx,7FFE2A573020h  
00007FFE2A98EAB7  mov         edx,5F5h  
00007FFE2A98EABC  call        00007FFE8A11F410  
00007FFE2A98EAC1  lea         rcx,[rax+40h]  
00007FFE2A98EAC5  mov         rdx,rsi  
00007FFE2A98EAC8  call        00007FFE8A40F7F0  
00007FFE2A98EACD  mov         rcx,qword ptr [rbp-30h]  
00007FFE2A98EAD1  call        00007FFE2A88E1C8  
00007FFE2A98EAD6  nop  
00007FFE2A98EAD7  lea         rsp,[rbp-20h]  
00007FFE2A98EADB  pop         rbx  
00007FFE2A98EADC  pop         rsi  
00007FFE2A98EADD  pop         rdi  
00007FFE2A98EADE  pop         r14  
00007FFE2A98EAE0  pop         rbp  
00007FFE2A98EAE1  ret  
    87:             ExecutionContextSwitcher ecsw = default(ExecutionContextSwitcher);
00007FFE2A98EAE2  push        rbp  
00007FFE2A98EAE3  push        r14  
00007FFE2A98EAE5  push        rdi  
00007FFE2A98EAE6  push        rsi  
00007FFE2A98EAE7  push        rbx  
00007FFE2A98EAE8  sub         rsp,30h  
00007FFE2A98EAEC  mov         rbp,qword ptr [rcx+20h]  
00007FFE2A98EAF0  mov         qword ptr [rsp+20h],rbp  
00007FFE2A98EAF5  lea         rbp,[rbp+60h]  
    96:             {
    97:                 // Note: we have a "catch" rather than a "finally" because we want
    98:                 // to stop the first pass of EH here.  That way we can restore the previous
    99:                 // context before any of our callers' EH filters run.  That means we need to 
   100:                 // end the scope separately in the non-exceptional case below.
   101:                 ecsw.Undo();
00007FFE2A98EAF9  lea         rcx,[rbp-30h]  
00007FFE2A98EAFD  call        00007FFE2A98D940  
   102:                 throw;
00007FFE2A98EB02  call        00007FFE8A1231F0  
   102:                 throw;
00007FFE2A98EB07  int         3  
--- No source file -------------------------------------------------------------
```
...ExecutionContext.Restore et al

```
--- I:\Work\GitHub\coreclr\src\mscorlib\src\System\Threading\ThreadPool.cs -----
  1309:             ccb = (state) => { ((QueueUserWorkItemCallbackDefaultContext)state).WaitCallback_Context(); };
00007FFE2A98ED10  sub         rsp,28h  
00007FFE2A98ED14  mov         rcx,rdx  
00007FFE2A98ED17  test        rcx,rcx  
00007FFE2A98ED1A  je          00007FFE2A98ED36  
00007FFE2A98ED1C  mov         rax,7FFE2A97D7C0h  
00007FFE2A98ED26  cmp         qword ptr [rcx],rax  
00007FFE2A98ED29  je          00007FFE2A98ED36  
00007FFE2A98ED2B  mov         rcx,rax  
00007FFE2A98ED2E  call        00007FFE8A40F5E0  
00007FFE2A98ED33  mov         rcx,rax  
00007FFE2A98ED36  mov         rax,7FFE2A98BE48h  
00007FFE2A98ED40  cmp         dword ptr [rcx],ecx  
00007FFE2A98ED42  add         rsp,28h  
00007FFE2A98ED46  jmp         rax  
--- No source file -------------------------------------------------------------
```
`private void WaitCallback_Context()`
```
--- I:\Work\GitHub\coreclr\src\mscorlib\src\System\Threading\ThreadPool.cs -----
  1315:             callback(state);
00007FFE2A98ED60  mov         qword ptr [rsp+8],rcx  
00007FFE2A98ED65  mov         rax,qword ptr [rcx+8]  
00007FFE2A98ED69  mov         rcx,qword ptr [rax+8]  
00007FFE2A98ED6D  mov         rdx,qword ptr [rsp+8]  
00007FFE2A98ED72  mov         rdx,qword ptr [rdx+10h]  
00007FFE2A98ED76  mov         rax,qword ptr [rax+18h]  
00007FFE2A98ED7A  jmp         rax  
--- No source file -------------------------------------------------------------
```
Then into my callback

Looks pretty trim?

**Comment by AlexGhiondea** [comment]
@benaadams do you have some performance numbers to show what this change will do?

**Comment by benaadams** [comment]
@AlexGhiondea these functions are called several million times a second so it is a hot path for performance.

However, I haven't managed to validate it yet. I disassembled and stepped through both versions (non-native image) and this version has slightly fewer instructions but there isn't that much change due to the magic of the jitter.

I'm going to have to compare and benchmark the native images which will take a bit longer.

**Comment by benaadams** [comment]
Closing, measured and integrated in #5943

**Pull request closed without merging**